yarn-project/ethereum/src/deploy_l1_contracts.ts

```diff
@@ -651,15 +652,63 @@ export async function deployL1Contract(
     const existing = await publicClient.getBytecode({ address });
 
     if (existing === undefined || existing === '0x') {
-      txHash = await walletClient.sendTransaction({ to: deployer, data: concatHex([salt, calldata]) });
-      logger?.verbose(`Deploying contract with salt ${salt} to address ${address} in tx ${txHash}`);
+      // Add gas estimation and price buffering for CREATE2 deployment
+      const deployData = encodeDeployData({ abi, bytecode, args });
+      const gasUtils = new GasUtils(publicClient, logger);
+      const gasEstimate = await gasUtils.estimateGas(() =>
+        publicClient.estimateGas({
+          account: walletClient.account?.address,
+          data: deployData,
+        }),
+      );
+      const gasPrice = await gasUtils.getGasPrice();
+
+      txHash = await walletClient.sendTransaction({
+        to: deployer,
+        data: concatHex([salt, calldata]),
+        gas: gasEstimate,
+        maxFeePerGas: gasPrice,
+      });
+      logger?.verbose(
+        `Deploying contract with salt ${salt} to address ${address} in tx ${txHash} (gas: ${gasEstimate}, price: ${gasPrice})`,
+      );
     } else {
       logger?.verbose(`Skipping existing deployment of contract with salt ${salt} to address ${address}`);
     }
   } else {
-    txHash = await walletClient.deployContract({ abi, bytecode, args });
-    logger?.verbose(`Deploying contract in tx ${txHash}`);
-    const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash, pollingInterval: 100 });
+    // Regular deployment path
+    const deployData = encodeDeployData({ abi, bytecode, args });
+    const gasUtils = new GasUtils(publicClient, logger);
+    const gasEstimate = await gasUtils.estimateGas(() =>
+      publicClient.estimateGas({
+        account: walletClient.account?.address,
+        data: deployData,
+      }),
+    );
+    const gasPrice = await gasUtils.getGasPrice();
+
+    txHash = await walletClient.deployContract({
+      abi,
+      bytecode,
+      args,
+      gas: gasEstimate,
+      maxFeePerGas: gasPrice,
+    });
+
+    // Monitor deployment transaction
+    await gasUtils.monitorTransaction(txHash, walletClient, {
+      to: '0x', // Contract creation
+      data: deployData,
+      nonce: await publicClient.getTransactionCount({ address: walletClient.account!.address }),
```

Comment:
You can run into a very nasty race condition here if the request gets routed to a different node than the one that received your tx above, so you end up getting an older nonce, which causes all replacements to fail.

I'm afraid we'll need to look into anvil internals to either get the nonce out of the `deployContract` result, or from within the walletClient itself. There is a `NonceManager` in viem which may have this info, though I'm not sure how to access it.